db/db_impl_compaction_flush.cc

```diff
@@ -200,6 +202,172 @@ Status DBImpl::FlushMemTableToOutputFile(
   return s;
 }
 
+/*
+ * Atomically flushes multiple column families.
+ *
+ * For each column family, all memtables with ID smaller than or equal to the
+ * specified ID will be flushed. Only after all column families finish flush
+ * will this function commit to MANIFEST. If any of the column families are not
+ * flushed successfully, this function does not have any side-effect on the
+ * state of the database.
+ */
+Status DBImpl::AtomicFlushMemTablesToOutputFiles(
+    const autovector<ColumnFamilyData*>& cfds,
+    const autovector<MutableCFOptions>& mutable_cf_options_list,
+    const autovector<uint64_t>& flush_memtable_ids, bool* made_progress,
+    JobContext* job_context, LogBuffer* log_buffer) {
+  mutex_.AssertHeld();
+#ifndef NDEBUG
+  for (const auto cfd : cfds) {
+    assert(cfd->imm()->NumNotFlushed() != 0);
+    assert(cfd->imm()->IsFlushPending());
+  }
+#endif /* !NDEBUG */
+
+  SequenceNumber earliest_write_conflict_snapshot;
+  std::vector<SequenceNumber> snapshot_seqs =
+      snapshots_.GetAll(&earliest_write_conflict_snapshot);
+
+  auto snapshot_checker = snapshot_checker_.get();
+  if (use_custom_gc_ && snapshot_checker == nullptr) {
```

Comment:
I know this is copied from elsewhere, but is it right? `DisableGCSnapshotChecker::IsInSnapshot` always returns false. `CompactionIterator` invokes `IsInSnapshot` to determine whether the next key is visible to `prev_snapshot`. If false, it treats the next key as if it's in the same snapshot as the current key. But this can cause the next key to be dropped, breaking the snapshot.

Reply:
cc @maysamyabandeh 

An example is like:

- SingleDelete "a" at seqnum 30
- Snapshot at seqnum 20
- Put "a" at seqnum 10

When CompactionIterator sees a SingleDelete, it checks whether both the current and next internal keys have the same user key and are in the same snapshot stripe. If so, they are dropped together.

I suspect (but haven't verified) that `DisableGCSnapshotChecker` always returning false can cause the SingleDelete and Put to be treated as part of the same snapshot stripe. Then they'd be mistakenly dropped together.

Reply:
Good catch. I'll look into it.

Reply:
@ajkr I checked the code and found out that this code path is executed only when `use_custom_gc_` is true. `use_custom_gc_` is used only by write prepared transactions.

Reply:
@ajkr I suggest that we address issue in a different PR.

Reply:
Yes, let's ask @maysamyabandeh to take a look if we both think it's a bug.

Reply:
SG